### DISCUSSION TITLE
Using conan for unit tests

### DISCUSSION THREAD

**Discussion by kammce** [friendlyanon/cmake-init#25] [Q&A] [answered]
Whats the preferred way to us conan libraries for unit tests. For example, my only dependencies for my library tests are the library itself and [boost::ut](https://conan.io/center/boost-ext-ut). 

My previous naive approach for this was just to have a `conanfile.txt`:
```
[requires]
libarmcortex/[>=0.0.1]
boost-ext-ut/1.1.8

[generators]
cmake_find_package_multi
```

And my `CMakeLists.txt` file looked like:

```cmake
cmake_minimum_required(VERSION 3.1...3.21)

project(test VERSION 0.0.1 LANGUAGES CXX)

list(APPEND CMAKE_MODULE_PATH ${CMAKE_BINARY_DIR})
list(APPEND CMAKE_PREFIX_PATH ${CMAKE_BINARY_DIR})

find_package(libarmcortex CONFIG REQUIRED)
find_package(ut CONFIG REQUIRED)

add_executable(${PROJECT_NAME}
    system_timer.test.cpp
    dwt_counter.test.cpp
    interrupt.test.cpp
    main.test.cpp)

target_compile_features(${PROJECT_NAME} PRIVATE cxx_std_20)
set_target_properties(${PROJECT_NAME} PROPERTIES CXX_EXTENSIONS OFF)
target_compile_definitions(${PROJECT_NAME} PRIVATE -DPLATFORM=test)
target_link_libraries(${PROJECT_NAME} PRIVATE
    libarmcortex::libarmcortex
    boost::ut)
```

I ran `conan install ..`, `cmake ..`, `make`, and `./test` from a build folder within the `tests/` directory and not from the top level. I'm not sure if that's the right approach though when it comes to using cmake-init.

**Comment by friendlyanon** [comment, marked as answer]
The [Conan example](https://github.com/friendlyanon/cmake-init-conan-example) from the README shows you how to integrate with Conan. It also shows you how to use a dev dependency, such as doctest. To costumize how Conan installs your dependencies, you can edit the conan.cmake file.

If you follow the example, then you have an experience similar to vcpkg:

```bash
cmake --preset=dev-unix # the preset also has the Conan code
cmake --build build/dev-unix
ctest --test-dir build/dev-unix
```

**Reply by kammce** [reply]
Would this be where the dependencies are listed? https://github.com/friendlyanon/cmake-init-conan-example/blob/master/cmake/conan.cmake#L45

Does that mean that using a conanfile.txt or conanfile.py is discouraged? And if so, how does something like `conan create .` get utilized in order to generate/locally install packages?

**Reply by friendlyanon** [reply]
It's not discouraged, but I have not yet found a way to express dev dependencies in conanfiles. Delegating the invocation of Conan to cmake-conan allows you to have requirements hidden behind an option. This sort of mimics vcpkg's "features" feature.
